File: leetcode/problems/problem-specific/16.3-sum-closest/16.3-sum-closest.go
package main

import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	if len(nums) < 3 {
		return 0
	}

	closest := math.MaxInt32
	sort.Ints(nums)
	for i := 0; i < len(nums)-1; i++ {
		left, right := i+1, len(nums)-1
		for left < right {
			totalSum := nums[i] + nums[left] + nums[right]
			if int(math.Abs(float64(totalSum-target))) < int(math.Abs(float64(closest-target))) {
				closest = totalSum
			}
			if totalSum == target {
				return totalSum
			} else if totalSum < target {
				left++
			} else {
				right--
			}
		}
	}

	return closest
}

File: leetcode/problems/language-specific/python/16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start

# * Two Pointers Solution | O(n^2) Time | O(1) Space
# * n -> The length of nums array


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums_len = len(nums)
        nums.sort()
        closest = float("inf")
        for i in range(nums_len - 2):
            left, right = i + 1, nums_len - 1
            while left < right:
                cur_sum = nums[i] + nums[left] + nums[right]
                if abs(target - cur_sum) < abs(target - closest):
                    closest = cur_sum
                if cur_sum == target:
                    return cur_sum
                elif cur_sum < target:
                    left += 1
                else:
                    right -= 1

        return closest


# @lc code=end
